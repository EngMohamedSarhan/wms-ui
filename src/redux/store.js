import { persistStore } from "redux-persist";

import { configureStore } from "@reduxjs/toolkit";

import { persistedReducer } from "./reducers/reducers";

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false,
    }),
});

export const persistor = persistStore(store);

export const settingsState = (state) => state.settingsReducer;
