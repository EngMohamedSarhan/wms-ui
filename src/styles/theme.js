import { createTheme } from "@mui/material";

const theme = createTheme({
  palette: {
    primary: {
      main: "#550016",
    },
    secondary: {
      main: "#1C0076",
    },
  },
});

export default theme;
