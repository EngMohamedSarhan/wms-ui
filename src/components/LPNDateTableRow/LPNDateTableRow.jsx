import { CheckOutlined, SaveOutlined } from "@mui/icons-material";
import {
  Fade,
  Grid,
  IconButton,
  TableCell,
  TextField,
  Tooltip,
} from "@mui/material";
import { DatePicker, MobileDatePicker } from "@mui/x-date-pickers";
import PropTypes from "prop-types";
import React, { useEffect, useRef, useState } from "react";
import { useDispatch, useSelector } from "react-redux";

import SquareButton from "../../components-styled/SquareButton/SquareButton";
import StyledTableRow from "../../components-styled/StyledTableRow/StyledTableRow";
import usePostAdjustOne from "../../hooks/usePostAdjustOne";
import { openNotification } from "../../redux/reducers/settingsSlice";
import { settingsState } from "../../redux/store";
import columns from "../../utilities/constants/lpn-date-table-columns";
import { getDateFromPicker } from "../../utilities/functions/date";
import { formatRow } from "../../utilities/functions/format";
import en from "../../utilities/json/en.json";

const LPNDateTableRow = ({
  unsavedRowsRef,
  row,
  index,
  onRowSave,
  saveAllCounter,
  setIsPageUpdated,
}) => {
  const dispatch = useDispatch();
  const isMftdDateChangedRef = useRef(false);
  const isExpiredDateChangedRef = useRef(false);
  const isCPDChangedRef = useRef(false);
  const { selectedWarehouse } = useSelector(settingsState);
  const { postAdjustOne } = usePostAdjustOne();
  const [manufactureDate, setManufactureDate] = useState(
    row.manufacturedDate || null
  );
  const [expirationDate, setExpirationDate] = useState(row.expirationDate || null);
  const [consumptionPriorityDate, setConsumptionPriorityDate] = useState(
    row.consumptionPriorityDate || null
  );
  const [isUpdated, setIsUpdated] = useState(false);

  const handleManufactureDateChange = (value) => {
    if (!value) {
      setManufactureDate(null);
      if (row.manufacturedDate) {
        setIsUpdated(true);
        isMftdDateChangedRef.current = true;
      }
      return;
    }

    const date = getDateFromPicker(value);

    if (date !== manufactureDate) {
      isMftdDateChangedRef.current = true;
      setIsUpdated(true);
      setManufactureDate(date);
    }
  };

  const handleExpirationDateChange = (value) => {
    if (!value) {
      setExpirationDate(null);
      if (row.expirationDate) {
        setIsUpdated(true);
        isExpiredDateChangedRef.current = true;
      }
      return;
    }

    const date = getDateFromPicker(value);

    if (date !== expirationDate) {
      isExpiredDateChangedRef.current = true;
      setIsUpdated(true);
      setExpirationDate(date);
    }
  };

  const handlePriortyDateChange = (value) => {
    if (!value) {
      setConsumptionPriorityDate(null);
      if (row.consumptionPriorityDate) {
        setIsUpdated(true);
        isCPDChangedRef.current = true;
      }
      return;
    }

    const date = typeof value === "string" ? value : getDateFromPicker(value);

    if (date !== consumptionPriorityDate) {
      isCPDChangedRef.current = true;
      setIsUpdated(true);
      setConsumptionPriorityDate(date);
    }
  };

  const handleSuggestedCPD = () => handlePriortyDateChange(row.suggestedCPD);

  const handlePostAdjustOne = () => {
    postAdjustOne(
      {
        lpnSingleAdjustRequest: formatRow(
          row,
          isMftdDateChangedRef.current,
          isExpiredDateChangedRef.current,
          isCPDChangedRef.current,
          manufactureDate,
          expirationDate,
          consumptionPriorityDate,
          selectedWarehouse
        ),
        location: selectedWarehouse,
      },
      ({ isSuccess, sourceContainerId, message }) => {
        setIsUpdated(false);
        dispatch(
          openNotification({
            title: en.details,
            message: [
              {
                message: `${sourceContainerId} ${
                  isSuccess ? en.saveSuccessMessage : message
                }`,
                isSuccess,
                status: isSuccess ? en.success : en.error,
              },
            ],
          })
        );

        if (isSuccess) {
          onRowSave(index);
        }
      }
    );
  };

  const handleReset = () => {
    setIsUpdated(false);
    setManufactureDate(row.manufacturedDate || null);
    setExpirationDate(row.expirationDate || null);
    setConsumptionPriorityDate(row.consumptionPriorityDate || null);
    isMftdDateChangedRef.current = false;
    isExpiredDateChangedRef.current = false;
    isCPDChangedRef.current = false;
  };

  const handleSave = () => {
    if (isUpdated) {
      handlePostAdjustOne();
    } else {
      dispatch(openNotification({ title: en.error, message: en.saveErrorMessage }));
    }
  };

  const renderColumns = () =>
    columns.map((column) => {
      const value = row[column.id];
      const renderChildren = (() => {
        switch (true) {
          case column.format && typeof value === "number":
            return column.format(value);
          case column.isManufactureDate:
            if (!row.trackManufacturingDate) {
              break;
            }

            return (
              <DatePicker
                value={manufactureDate}
                inputFormat="MM/DD/YYYY"
                onChange={handleManufactureDateChange}
                renderInput={(params) => <TextField size="small" {...params} />}
              />
            );
          case column.isExpirationDate:
            if (!row.trackExpiryDate) {
              break;
            }

            return (
              <DatePicker
                value={expirationDate}
                inputFormat="MM/DD/YYYY"
                onChange={handleExpirationDateChange}
                renderInput={(params) => <TextField size="small" {...params} />}
              />
            );
          case column.isPriorityDate:
            return (
              <Grid container alignItems="center">
                <Grid item xs={2}>
                  {row.suggestedCPD ? (
                    <Fade in={consumptionPriorityDate !== row.suggestedCPD}>
                      <Tooltip title={en.suggestedCpdTooltip}>
                        <IconButton sx={{ p: 0 }} onClick={handleSuggestedCPD}>
                          <CheckOutlined color="success" />
                        </IconButton>
                      </Tooltip>
                    </Fade>
                  ) : null}
                </Grid>
                <Grid item xs={10}>
                  <DatePicker
                    value={consumptionPriorityDate}
                    inputFormat="MM/DD/YYYY"
                    onChange={handlePriortyDateChange}
                    renderInput={(params) => <TextField size="small" {...params} />}
                  />
                </Grid>
              </Grid>
            );
          case column.isSuggestedCPD:
            if (!value) {
              break;
            }

            return (
              <MobileDatePicker
                readOnly
                value={value}
                inputFormat="MM/DD/YYYY"
                renderInput={(params) => <TextField size="small" {...params} />}
              />
            );
          case column.isSave:
            return (
              <SquareButton
                variant="contained"
                color="primary"
                onClick={handleSave}
                sx={{
                  mr: 1,
                  backgroundColor: "#fff",
                  ["&:hover"]: {
                    backgroundColor: "primary.main",
                    color: "#fff",
                    [".MuiSvgIcon-root"]: {
                      color: "#fff",
                    },
                  },
                }}
              >
                <SaveOutlined color="primary" />
              </SquareButton>
            );
          default:
            return value;
        }
      })();

      return (
        <TableCell
          key={column.id}
          align={column.align}
          sx={{ p: 0.5, pl: 1, pr: 1, ...column.sx }}
        >
          {renderChildren}
        </TableCell>
      );
    });

  useEffect(() => {
    handleReset();
  }, [saveAllCounter]);

  useEffect(() => {
    if (isUpdated) {
      unsavedRowsRef.current[index] = formatRow(
        row,
        isMftdDateChangedRef.current,
        isExpiredDateChangedRef.current,
        isCPDChangedRef.current,
        manufactureDate,
        expirationDate,
        consumptionPriorityDate,
        selectedWarehouse
      );
      setIsPageUpdated(true);
    }
  }, [expirationDate, manufactureDate, consumptionPriorityDate, isUpdated]);

  return (
    <StyledTableRow
      hover={!isUpdated}
      change={isUpdated}
      role="checkbox"
      tabIndex={-1}
      key={row.code}
    >
      {renderColumns()}
    </StyledTableRow>
  );
};

LPNDateTableRow.propTypes = {
  row: PropTypes.object.isRequired,
  index: PropTypes.number.isRequired,
  onRowSave: PropTypes.func.isRequired,
  unsavedRowsRef: PropTypes.object.isRequired,
  setIsPageUpdated: PropTypes.func.isRequired,
  saveAllCounter: PropTypes.number.isRequired,
};

export default LPNDateTableRow;
