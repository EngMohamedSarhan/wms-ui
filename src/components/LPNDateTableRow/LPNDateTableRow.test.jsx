/* eslint-disable no-undef */
import { LocalizationProvider } from "@mui/x-date-pickers";
import { AdapterDayjs } from "@mui/x-date-pickers/AdapterDayjs";
import { fireEvent, render, screen } from "@testing-library/react";
import { React } from "react";
import { Provider } from "react-redux";
import ShallowRenderer from "react-shallow-renderer";

import { store } from "../../redux/store";
import LPNDateTableRow from "./LPNDateTableRow";
import theme from "../../styles/theme";
import { ThemeProvider } from "@mui/material";

const rowData = {
  id: "0",
  itemId: "532432",
  itemDescription: "243 SA",
  ilpnId: "2222X",
  lpnStatus: "Not Allocated",
  locationId: "06203A",
  lpnQuantityInCases: 13,
  reason: "EXP DATE ITEM MISSING EXPDATE",
  dateCode: "E",
  trackManufacturingDate: 1,
  trackExpiryDate: 1,
  mfgDateDiff: 8,
  manufacturedDate: "2021-06-17T",
  expirationDate: "2022-05-13T",
  consumptionPriorityDate: "2022-06-17T",
};

describe("LPNDateTableRow Component", () =>
  test("LPNDateTableRow action", () => {
    const onRowUpdate = jest.fn();
    const onRowSave = jest.fn();
    const setIsPageUpdated = jest.fn();
    const onRowSaveFailed = jest.fn();
    const onRowValidation = jest.fn();
    const { container } = render(
      <Provider store={store}>
        <LocalizationProvider dateAdapter={AdapterDayjs}>
          <LPNDateTableRow
            row={rowData}
            index={0}
            unsavedRowsRef={{}}
            onRowValidation={onRowValidation}
            onRowUpdate={onRowUpdate}
            onRowSave={onRowSave}
            onRowSaveFailed={onRowSaveFailed}
            setIsPageUpdated={setIsPageUpdated}
            saveAllCounter={0}
          />
        </LocalizationProvider>
      </Provider>
    );

    const saveButton = container.getElementsByTagName("button")[0];
    fireEvent.click(saveButton);
    expect(onRowUpdate).not.toBeCalled();
  }));

describe("LPNDateTableRow Component", () =>
  test("LPNDateTableRow snapshot", () => {
    const onRowUpdate = jest.fn();
    const onRowSave = jest.fn();
    const onRowSaveFailed = jest.fn();
    const setIsUpdated = jest.fn();
    const onRowValidation = jest.fn();

    const renderer = new ShallowRenderer();
    renderer.render(
      <Provider store={store}>
        <LocalizationProvider dateAdapter={AdapterDayjs}>
          <LPNDateTableRow
            row={rowData}
            index={0}
            unsavedRowsRef={{}}
            onRowValidation={onRowValidation}
            onRowUpdate={onRowUpdate}
            saveAllCounter={0}
            setIsPageUpdated={setIsUpdated}
            onRowSave={onRowSave}
            onRowSaveFailed={onRowSaveFailed}
          />
        </LocalizationProvider>
      </Provider>
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  }));

describe("LPNDateTableRow Component", () =>
  test("LPNDateTableRow action", () => {
    const onRowUpdate = jest.fn();
    const onRowSave = jest.fn();
    const onRowSaveFailed = jest.fn();
    const ref = { current: {} };
    const setIsUpdated = jest.fn();
    const onRowValidation = jest.fn();

    render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <LocalizationProvider dateAdapter={AdapterDayjs}>
            <LPNDateTableRow
              row={rowData}
              index={0}
              unsavedRowsRef={ref}
              onRowValidation={onRowValidation}
              onRowUpdate={onRowUpdate}
              onRowSave={onRowSave}
              onRowSaveFailed={onRowSaveFailed}
              saveAllCounter={0}
              setIsPageUpdated={setIsUpdated}
            />
          </LocalizationProvider>
        </ThemeProvider>
      </Provider>
    );
    const mfgDatePicker = screen.getAllByPlaceholderText("mm/dd/yyyy")[0];
    const expDatePicker = screen.getAllByPlaceholderText("mm/dd/yyyy")[1];
    const conPrioDatePicker = screen.getAllByPlaceholderText("mm/dd/yyyy")[2];

    expect(mfgDatePicker.value).toEqual("06/17/2021");

    fireEvent.change(mfgDatePicker, { target: { value: "08/20/2022" } });
    expect(mfgDatePicker.value).toEqual("08/20/2022");

    fireEvent.change(expDatePicker, { target: { value: "08/20/2022" } });
    expect(expDatePicker.value).toEqual("08/20/2022");

    fireEvent.change(conPrioDatePicker, { target: { value: "08/20/2022" } });
    expect(conPrioDatePicker.value).toEqual("08/20/2022");
  }));
