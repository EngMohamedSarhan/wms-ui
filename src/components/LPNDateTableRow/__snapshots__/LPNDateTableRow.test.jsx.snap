// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LPNDateTableRow Component LPNDateTableRow snapshot 1`] = `
<Context.Provider
  value={
    Object {
      "getServerState": undefined,
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <LocalizationProvider
    dateAdapter={[Function]}
  >
    <LPNDateTableRow
      index={0}
      onRowSave={[MockFunction]}
      row={Object {}}
      saveAllCounter={0}
      setIsPageUpdated={[MockFunction]}
      unsavedRowsRef={Object {}}
    />
  </LocalizationProvider>
</Context.Provider>
`;
